libnestutil/propagator_stability.h

```diff
@@ -23,9 +23,154 @@
+ * propagator_31 and propagator_32 are calculated in `evaluate( h )`.
+ *
+ * For details, please see doc/userdoc/model_details/IAF_neurons_singularity.ipynb.
+ */
+class Propagator
+{
+public:
+  /**
+   * Empty constructor needed for initialization of buffers.
+   */
+  Propagator();
+
+  /**
+   * @param tau_syn Time constant of synaptic current in ms
+   * @param tau_m Membrane time constant in ms
+   * @param c_m Membrane capacitance in pF
+   */
+  Propagator( double tau_syn, double tau_m, double c_m );
+
+protected:
+  /**
+   * Calculate propagator 32 and return value along with exponentials.
+   *
+   * Exponentials are returned so they can be used directly when calculating P31.
+   *
+   * @param h time step
+   *
+   * @returns tuple with propagator 32, exp( -h / tau_syn_ ), expm1( -h / tau_m_ + h / tau_syn_ )
+   * and exp( -h / tau_m_ ), all as doubles
+   */
+  std::tuple< double, double, double, double > evaluate_P32_( double h ) const;
+
+  double tau_syn_; //!< Time constant of synaptic current in ms
+  double tau_m_;   //!< Membrane time constant in ms
+  double c_m_;     //!< Membrane capacitance in pF
+
+  double alpha_; //!< 1/(c*tau*tau) * (tau_syn - tau)
+  double beta_;  //!< tau_syn * tau/(tau - tau_syn)
+  double gamma_; //!< beta_/c
+};
+
+
+/**
+ * Propagator class for handling similar tau_m and tau_syn_* time constants for
+ * models with exponential postsynaptic currents.
+ *
+ * propagator_32 is calculated in `evaluate( h )` and returned as a double.
+ */
+class PropagatorExp : public Propagator
+{
+public:
+  /**
+   * Empty constructor needed for initialization of buffers.
+   */
+  PropagatorExp();
+
+  /**
+   * @param tau_syn Time constant of synaptic current in ms
+   * @param tau_m Membrane time constant in ms
+   * @param c_m Membrane capacitance in pF
+   */
+  PropagatorExp( double tau_syn, double tau_m, double c_m );
+
+  /**
+   * Calculate propagator 32.
+   *
+   * @param h time step
+   *
+   * @returns propagator 32 as a double
+   */
+  double evaluate( double h ) const;
+};
+
+/**
+ * Propagator class for handling similar tau_m and tau_syn_* time constants for
+ * models with postsynaptic currents modeled as an alpha current.
+ *
+ * propagator_31 and propagator_32 are calculated in `evaluate( h )` and returned
+ * as a tuple, where P31 is the first variable.
+ */
+class PropagatorAlpha : public Propagator
+{
+public:
+  /**
+   * Empty constructor needed for initialization of buffers.
+   */
+  PropagatorAlpha();
+
+  /**
+   * @param tau_syn Time constant of synaptic current in ms
+   * @param tau Membrane time constant in ms
+   * @param Membrane capacitance in pF
+   */
+  PropagatorAlpha( double tau_syn, double tau_m, double c_m );
+
+  /**
+   * Calculate propagators 31 and 32.
+   *
+   * @param h time step
+   *
+   * @returns tuple containing propagators 31 and 32
+   */
+  std::tuple< double, double > evaluate( double h ) const;
+};
+
+inline std::tuple< double, double, double, double >
```

Comment:
This method is so large that I see no point in inlining, so I suggest to move it to the cpp-file.